pennylane/interfaces/batch/jax.py

```diff
@@ -0,0 +1,257 @@
+                if len(p) == 1:
+                    res[-1] = res[-1][0]
+
+            # Unwrap partial results into arrays if need be
+            need_unwrapping = any(r.ndim != 0 for r in res)
+            if need_unwrapping:
+                unwrapped_res = []
+                for r in res:
+                    if r.ndim != 0:
+                        r = [jnp.array(p) for p in r]
+                    unwrapped_res.append(r)
+
+                res = unwrapped_res
+
+            return (tuple(res),)
+
+        # Gradient function is a device method.
+        with qml.tape.Unwrap(*tapes):
+            jacs = gradient_fn(tapes, **gradient_kwargs)
+
+        vjps = [qml.gradients.compute_vjp(d, jac) for d, jac in zip(g, jacs)]
+        res = [[jnp.array(p) for p in v] for v in vjps]
+        return (tuple(res),)
+
+    wrapped_exec.defvjp(wrapped_exec_fwd, wrapped_exec_bwd)
+    return wrapped_exec(params)
+
+
+# The execute function in forward mode
+def _execute_with_fwd(
+    params,
+    tapes=None,
+    device=None,
+    execute_fn=None,
+    gradient_kwargs=None,
+    _n=1,
+):  # pylint: disable=dangerous-default-value,unused-argument
+
+    # Only have scalar outputs
+    total_size = len(tapes)
+
+    @jax.custom_vjp
+    def wrapped_exec(params):
+        def wrapper(p):
+            """Compute the forward pass by returning the jacobian too."""
+            new_tapes = []
+
+            for t, a in zip(tapes, p):
+                new_tapes.append(t.copy(copy_operations=True))
+                new_tapes[-1].set_parameters(a)
+
+            res, jacs = execute_fn(new_tapes, **gradient_kwargs)
+
+            # On the forward execution return the jacobian too
+            return np.stack(res), jacs
+
+        jacobian_shape = [jax.ShapeDtypeStruct((1, len(p)), dtype) for p in params]
+        res, jacs = host_callback.call(
+            wrapper,
+            params,
+            result_shape=tuple([jax.ShapeDtypeStruct((total_size, 1), dtype), jacobian_shape]),
+        )
+        return res, jacs
+
+    def wrapped_exec_fwd(params):
+        res, jacs = wrapped_exec(params)
+        return res, tuple([jacs, params])
+
+    def wrapped_exec_bwd(params, g):
+
+        # Use the jacobian that was computed on the forward pass
+        jacs, params = params
+        res_jacs = []
+        for j in jacs:
+            this_j = []
+            for i in range(j.shape[1]):
+                this_j.append(j[0, i])
+            res_jacs.append(this_j)
+        return tuple([tuple(res_jacs)])
+
+    wrapped_exec.defvjp(wrapped_exec_fwd, wrapped_exec_bwd)
+    res = wrapped_exec(params)
+
+    tracing = any(isinstance(r, jax.interpreters.ad.JVPTracer) for r in res)
+
+    # When there are no tracers (not differentiating), we have the result of
+    # the forward pass and the jacobian, but only need the result of the
+    # forward pass
+    if len(res) == 2 and not tracing:
+        res = res[0]
+
+    return res

```

Comment:
When there _is_ tracing, how come you don't need to index into `res`?

Reply:
My understanding is that in such cases JAX produces the:
* the forward pass value by retrieving the first element of the output of `wrapped_exec_fwd`
* the backward pass value by retrieving the first element of the output of `wrapped_exec_bwd`

```python
from jax import custom_vjp
import jax.numpy as jnp
import jax

def func(x, y):

    @custom_vjp
    def f(x, y):
        return jnp.sin(x) * y, 0.123456

    def f_fwd(x, y):
    # Returns primal output and residuals to be used in backward pass by f_bwd.
        val, jac = f(x, y)
        return val, (jac, (x, y))

    def f_bwd(res, g):
        jac, params = res # Gets residuals computed in f_fwd
        return (jac, 0.0)

    f.defvjp(f_fwd, f_bwd)
    res = f(x,y)
    print('Output of f call: ', res)
    return res

print('Func call (no tracing or differentiation): ', func(0.3, 0.5), '\n\n')
print('Grad of func call: ', jax.grad(func)(0.3, 0.5))
```
```
Output of f call:  (DeviceArray(0.14776011, dtype=float32, weak_type=True), 0.123456)
Func call (no tracing or differentiation):  (DeviceArray(0.14776011, dtype=float32, weak_type=True), 0.123456) 


Output of f call:  Traced<ConcreteArray(0.14776010811328888, dtype=float32)>with<JVPTrace(level=2/0)>
  with primal = DeviceArray(0.14776011, dtype=float32, weak_type=True)
       tangent = Traced<ShapedArray(float32[]):JaxprTrace(level=1/0)>
Grad of func call:  0.123456
```
The second case of printing `Output of f call:` only contains the traced array.